### PR TITLE
[ECAM] Add Autopilot & A/THR warning and caution when disconnect

### LINKED ISSUE #1503: [REQUEST] Autopilot disconnect sound.
**Is your feature request related to a problem? Please describe.**

When disconnect autothrust, there is a no sound and no warning about disconnect autothrust

If we press autothrottle disconnect button on thurst lever's red button, a320neo should beep 'warning sound' once, and beep 3 times. 

**Describe the feature you'd like to see implemented**

**References**

https://youtu.be/Vq8Lj0pp39o?t=417
you can watch when disconnect autothurst

**Additional context**

Discord username (if different from GitHub):

### PR DESCRIPTION
**Pull request by Kimbyeoungjang**
Fixes #1503

## Summary of Changes

as a title, when disconnect autopilot or autothrottle, ECAM show warning or caution and light's on and off

## References

image

## Additional context

Discord username (if different from GitHub):

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Kimbyeoungjang** [comment]
work finished but need to review.
warning message is working well but when other warning(like eng fail etc..) is on, this logic can off other warning light. should change that or remove warning light logic only.

**Comment by Awemeter** [comment]
Name: Awemeter#7919
Date of testing: 11/19/2020
Version of the sim: 1.10.11.0
PR Tested: #2013 
Object of test: ECAM Page

Testing technique: Did a flight.
Steps you've made to test the Pull Request:
1. Pushed back from the gate at Kord and taxied to rwy 10L
2. took off rwy 10L flew heading 100 and enabled Auto Pilot. Then I went direct bucke and climbed to 10,000. After I reached 10,000 climbed to fl280
3. pulled standard at fl220
4. Descended to fl210 26nm away from bucke with open decent.
5. Descended to 17,000 4nm away from tindr with open decent and selected 280 speed.
6.14nm away from smms descended to 12,000 open descent and selected 260kts.
7. 17 nm away from dreme descended to 9,000 with open decent and selected speed 230kts
8. set 3007 altimeter at 10,000ft
9. descended to 6,000 13nm away from Valli with open decent.
10. Aircraft went direct to the runway so I flew heading 130.
11. selected speed 200 kts
12. went flap one 14nm away from runway 22L and activated ils
13. gear down and flap 3 on 3nm final
14. Unstable ils approach disengaged ils and went manual. Got warning on ecam in red when disengaged ap. 
15. Landed on runway 22L at kdtw -175fpm.
16. taxied to the gate and shut down engines.



Flight notes

Startup: Everything went as expected with the start-up.

Taxi/Takeoff : Everything Normal.

Cruise: Everything went normal

Descent/Landing: When on gs for ils pitched down aggressively to catch gs which made me over speed. Ils descended to low.

Shutdown: Went as expected.

Issues (If Any): 
Overall Rating: Good
Conclusions of the testing: Everything went well except ils approach.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 1]
```diff
@@ -975,6 +1013,22 @@ var A320_Neo_UpperECAM;
...
+                        isActive: () => {
+                            return (this.getAPWarning(3) == true);
+                        }
+                    },
+                    {
+                        message: "A/THR OFF",
+                        style: "fail-2",
+                        important: true,
+                        isActive: () => {
+                            return (this.getATHRWarning(3) == true);
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:1029]: `== true` is unnecessary here. You can remove it along with the parentheses surrounding the whole thing.

Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js]: i'll change that

**Comment by tjburni** [comment]
Category: QA Tester
Name: tjburni # 0409
Date of testing: Current date 21.11.2020
Version of the sim:1.10.11.0
PR Tested: #2013
Object of test:Autopilot & A/THR warning and caution when disconnect 

Testing technique:
1.Started at Gate 
2.Climed to 30000 ft 
3.Disconnected Autothrottle
4.Disconnected AP 
5.looked at Ecam for the Warnings

Issues:None 
Overall Rating: Good
Conclusions of the testing:Works Good. No Problems

**Comment by Benjozork** [comment]
Now only needs accuracy check.

**Comment by St54Kevin** [comment]
Tried an Accuracy check. But this PR is not working. Spawned in the Air in cruise and pressed the A/THR disco push button once. A/THR disconnected, but no ECAM Warning and sound, just the ECAM message. Here is a reference you should consider having a look at @Kimbyeoungjang 
image

**Comment by St54Kevin** [comment]
Also here is the same reference table for the AP:
image

**Comment by Kimbyeoungjang** [comment]
> Tried an Accuracy check. But this PR is not working. Spawned in the Air in cruise and pressed the A/THR disco push button once. A/THR disconnected, but no ECAM Warning and sound, just the ECAM message. Also pushing the button again did not delete the message. Here is a reference you should consider having a look at @Kimbyeoungjang
> image

I check that problems.

IDK why sound.xml is not working(was working but not now. I was recoded on my youtube, sound system was working)

Also i’ll add warning cancel system
Change this to draft...

**Comment by Kimbyeoungjang** [comment]
i will fix that when #2077 is solved. at this time, i can't touch sound problem.
i'll add button function that can cancel

**Comment by Highflyer2010** [comment]
Category: QA Tester
Name: Discord Highflyer2010 #8932
Date of testing: Current date 09/12/2020
Version of the sim: 1.11.7.0
PR Tested: #2013 
Object of test: Accuracy Check on A/THR Disconnect

Testing technique: took of from bejing with engines running, climbed to 10'00ft made sure aircraft was stable before testing.

Steps you've made to test the Pull Request:
1. started with engine running at ZBAA
2. Took off to 10'000ft (Aircraft stable)
3. disconnected A/THR - Nothing happened regarding sounds or warnings as shown in first video
4. Enabled A/THR because the aircraft started to speed up
5. Brought the throttle back to see if i get anything warning or sounds (Nothing happened) as shown in second video
6. then dissconnected A/THR - Again nothing happened

Flight notes (If required):

Issues (If Any): 
Overall Rating: Poor
Conclusions of the testing: I never got any warning sounds or messaged on ECAM which i should of done.

I will do another test after work starting from C&D to see if i get the same outcome 

1st Video
https://files.fm/f/889tzgvny

2nd Vid
https://files.fm/f/hhnarkkhz

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 2]
```diff
@@ -1004,6 +1042,22 @@ var A320_Neo_UpperECAM;
                         important: true,
                         isActive: () => this.showLandingInhibit,
                     },
+                    {
+                        message: "AP OFF",
+                        style: "fail-3",
+                        important: true,
+                        isActive: () => {
+                            return (this.getAPWarning(3));
```
Comment (beheh) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:1050]: Don't call functions that change the state or SimVars in isActive. IsActive may be called multiple times at the same time in future.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 3]
```diff
@@ -347,6 +347,44 @@ var A320_Neo_UpperECAM;
                 }
             }
         }
+        getAPWarning(time) {
+            const count = time * 1000;
+            if (this.getCachedSimVar("AUTOPILOT MASTER", "Bool") == false) {
+                const timer = setTimeout(() => {
```
Comment (beheh) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:353]: Don't use setTimeout. You always want to use _deltaTime as that's provided by the sim and can deal with variable FPS much better and more accurately.

**Review by beheh — CHANGES_REQUESTED** [review]
Quite a few issues with this PR: You should not be using setTimeout, and the way you use IsActive to change SimVars is prone to future breakage.

I would suggest moving your SimVar and timing logic into a separate function in this file, similar to say `updateInhibitMessages`, called by `update`. You'll then need some kind of timer to track the time differences, or you just use the NXLogic nodes for that. You can then set a variable that the `isActive` functions check in a read-only fashion.

**Comment by Kimbyeoungjang** [comment]
> Category: QA Tester
> Name: Discord Highflyer2010 #8932
> Date of testing: Current date 09/12/2020
> Version of the sim: 1.11.7.0
> PR Tested: #2013
> Object of test: Accuracy Check on A/THR Disconnect
> 
> Testing technique: took of from bejing with engines running, climbed to 10'00ft made sure aircraft was stable before testing.
> 
> Steps you've made to test the Pull Request:
> 
> 1. started with engine running at ZBAA
> 2. Took off to 10'000ft (Aircraft stable)
> 3. disconnected A/THR - Nothing happened regarding sounds or warnings as shown in first video
> 4. Enabled A/THR because the aircraft started to speed up
> 5. Brought the throttle back to see if i get anything warning or sounds (Nothing happened) as shown in second video
> 6. then dissconnected A/THR - Again nothing happened
> 
> Flight notes (If required):
> 
> Issues (If Any):
> Overall Rating: Poor
> Conclusions of the testing: I never got any warning sounds or messaged on ECAM which i should of done.
> 
> I will do another test after work starting from C&D to see if i get the same outcome
> 
> 1st Video
> https://files.fm/f/889tzgvny
> 
> 2nd Vid
> https://files.fm/f/hhnarkkhz

A.... this is draft(still in progress) so you don’t have to test this.
Thx for your feedback. Also sound is not working on current master. Some logic is changed and sound is not working(was working...)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js (hunk 3 above)
Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:353]: > There's e a few issues with this PR: You should not be using setTimeout, and the way you use IsActive to change SimVars is prone to future breakage.
> 
> I would suggest moving your SimVar and timing logic into a separate function in this file, similar to say `updateInhibitMessages`, called by `update`. You'll then need some kind of timer to track the time differences, or you just use the NXLogic nodes for that. You can then set a variable that the `isActive` functions check in a read-only fashion.

Thx for your feedback. Currently, this is problem that calling caution and warning. I’ll build another logic for that.

**Comment by Highflyer2010** [comment]
@Kimbyeoungjang sorry buddy didn't realise was just drinking my first cup of coffee when i was testing lol

**Comment by Kimbyeoungjang** [comment]
> @Kimbyeoungjang sorry buddy didn't realise was just drinking my first cup of coffee when i was testing lol

It’s okay :) when i finish this pr, then test this. Thank you for your feedback :)

**Comment by ktechmidas** [comment]
If you're below the CLB detent but still in A/THR there's also a warning on the real aircraft, any chance that can be added here as well?

**Comment by Kimbyeoungjang** [comment]
> If you're below the CLB detent but still in A/THR there's also a warning on the real aircraft, any chance that can be added here as well?

do you have reference about that? i'll add that :)

currently i have some important exam so my pr is late some time. until december 16? maybe?

**Comment by ktechmidas** [comment]
> do you have reference about that? i'll add that :)

Unfortunately I don't have the book to hand, but the message is AUTO FLT A/THR LIMITED. I can show some screenshots from the FSL A320 though?

**Comment by Kimbyeoungjang** [comment]
> > do you have reference about that? i'll add that :)
> 
> Unfortunately I don't have the book to hand, but the message is AUTO FLT A/THR LIMITED. I can show some screenshots from the FSL A320 though?

sure :) that might help us

**Comment by Kimbyeoungjang** [comment]
I'm reworking with new pr, closed.

**Pull request closed without merging**